pkg/status/reporter.go

```diff
@@ -0,0 +1,209 @@
+	r := &reporter{
+		name:       name,
+		client:     client,
+		ch:         make(chan *apiv3.CalicoNodeStatus, 10),
+		populators: populators,
+		done:       make(chan struct{}),
+	}
+
+	r.checkAndUpdateTicker(intervalInSeconds)
+
+	go r.run()
+	return r
+}
+
+// Check and set new ticker for the reporter.
+// Make sure stop the old one first to GC old ticker.
+func (r reporter) checkAndUpdateTicker(interval int) {
+	if r.interval == interval {
+		// no update needed.
+		return
+	}
+	if r.ticker != nil {
+		r.ticker.Stop()
+	}
+	r.interval = interval
+	r.ticker = time.NewTicker(time.Duration(interval) * time.Second)
+}
+
+// Cleanup resources owned by this reporter.
+func (r reporter) cleanup() {
+	r.ticker.Stop()
+}
+
+// KillAndWait sends done signal to reporter goroutine and wait until the
+// goroutine of the reporter terminated.
+func (r reporter) KillAndWait() {
+	r.done <- struct{}{}
+	<-r.term
+}
+
+// Called when the caller needs to send a new version of request.
+func (r reporter) RequestUpdate(request *apiv3.CalicoNodeStatus) {
+	r.ch <- request
+}
+
+// run is the main reporting loop, it loops until done.
+func (r reporter) run() {
+	runImmediately := make(chan struct{})
+
+	for {
+		select {
+		case latest := <-r.ch:
+			// Received an update of node status resource.
+			if latest.Name != r.name {
+				log.Warningf("node status reporter (%s) receive request with different name (%s), ignore it", r.name, latest.Name)
+			} else {
+				r.status = latest
+				r.checkAndUpdateTicker(latest.Spec.UpdateIntervalInSeconds)
+				// kick start node status update
+				runImmediately <- struct{}{}
+			}
+		case <-runImmediately:
+		case <-r.ticker.C:
+			if r.status == nil {
+				log.Warningf("No request found for node status reporter (%s)", r.name)
+			} else {
+				r.reportStatus()
+			}
+		case <-r.done:
+			r.cleanup()
+			r.term <- struct{}{}
+			return
+		}
+	}
+}
+
+// reportStatus queries Bird or other components and update node status resource.
+func (r reporter) reportStatus() {
+	var err error
+	// The idea here is that we either update everything successfully or we update nothing.
+
+	// Make a local copy first.
+	status := *r.status
+
+	needUpdate := false
+	for _, ipv := range []BirdConnType{BirdConnTypeV4, BirdConnTypeV6} {
+		failed := false
+		// Populate status from registered populators.
+		for _, class := range r.status.Spec.Classes {
+			if p, ok := r.populators[ipv][class]; ok {
+				err := p.Populate(&status)
+				if err != nil {
+					// If we hit error on one BirdConnType, continue with other BirdConnTypes.
+					log.WithError(err).Errorf("failed to populate status for ipv%s class %s", string(ipv), string(class))
+					failed = true
+					break
+				}
+			} else {
+				log.Warningf("Wrong class (%s) requested for node status reporter (%s)", class, r.name)
+			}
+		}
+
+		if !failed {
+			needUpdate = true
+		}
+	}
+
+	if !needUpdate {
```

Comment:
See my suggestion above about simply erroring if any of the populators fail - I think that is my preferred approach rather than handling partial failures, since inconsistent state in the API will be confusing for users. 

However, if we do for some reason need to keep this, can't we just remove `needUpdate` and turn this check into `if !failed`?